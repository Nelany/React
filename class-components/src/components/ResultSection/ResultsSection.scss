@import '../../utils/variables.scss';

.results-section {
  flex-grow: 1;
  background-color: $light;
}

.results-section.dark {
  background-color: $black;
}