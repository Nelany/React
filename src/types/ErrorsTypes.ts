export interface Errors {
  name?: string;
  age?: string;
  email?: string;
  password?: string;
  confirmPassword?: string;
  gender?: string;
  terms?: string;
  picture?: string;
  country?: string;
}
