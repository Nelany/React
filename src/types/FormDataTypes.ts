export interface FormData {
  name: string;
  age: string;
  email: string;
  password: string;
  confirmPassword: string;
  gender: string;
  terms: boolean;
  picture: string | null;
  country: string;
}
