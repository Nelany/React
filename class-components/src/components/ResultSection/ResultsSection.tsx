import { ResultsItem } from '../ResultsItem/ResultsItem';
import { CharacterResponse } from '../../types/types';
import { Loader } from '../Loader/Loader';
import { Pagination } from '../Pagination/Pagination';
import { Dispatch, SetStateAction } from 'react';
import './ResultsSection.scss';
import { useTheme } from '../../hooks/useTheme';

interface Props {
  characterResponse: CharacterResponse | null;
  isLoading: boolean;
  ifNextPage: boolean;
  setIfReturnToRickNMorty?: Dispatch<SetStateAction<boolean>>;
}

export const ResultsSection = ({
  characterResponse,
  isLoading,
  ifNextPage,
  setIfReturnToRickNMorty,
}: Props) => {
  const { theme } = useTheme();

  return (
    <div className={`section results-section ${theme}`}>
      <h3>Results:</h3>
      <Loader
        setIfReturnToRickNMorty={setIfReturnToRickNMorty}
        isLoading={isLoading}
        response={characterResponse || {}}
      />

      {!isLoading && characterResponse?.results && (
        <>
          {characterResponse.results.map((character) => (
            <ResultsItem
              key={character.id}
              name={character.name}
              character={character}
            />
          ))}
          <Pagination ifNextPage={ifNextPage} />
        </>
      )}
    </div>
  );
};
