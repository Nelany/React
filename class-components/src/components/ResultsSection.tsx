import { ResultsItem } from './ResultsItem';
import { CharacterResponse } from '../types/types';

interface Props {
  characterResponse: CharacterResponse | null;
  isLoading: boolean;
}

export const ResultsSection = ({ characterResponse, isLoading }: Props) => {
  return (
    <div className="section results-section">
      <h3>Results:</h3>
      {isLoading && (
        <div className="spinner-container">
          <img className="spinner" src="./spinner.png" alt="Loading..." />
        </div>
      )}
      {!isLoading && characterResponse?.error && (
        <h4>{`${characterResponse.error}!`}</h4>
      )}
      {!isLoading &&
        characterResponse?.results &&
        characterResponse.results.map((character) => (
          <ResultsItem
            key={character.id}
            name={character.name}
            character={character}
          />
        ))}
    </div>
  );
};
