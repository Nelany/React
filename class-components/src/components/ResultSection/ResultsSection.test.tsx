import { render, screen } from '@testing-library/react';
import { ResultsSection } from './ResultsSection';
import { CharacterResponse } from '../../types/types';
import '@testing-library/jest-dom';
import { MemoryRouter } from 'react-router-dom';

const mockCharacterResponse: CharacterResponse = {
  info: {
    count: 2,
    pages: 1,
    next: null,
    prev: null,
  },
  results: [
    {
      id: 1,
      name: 'Rick Sanchez',
      status: 'Alive',
      species: 'Human',
      type: '',
      gender: 'Male',
      origin: {
        name: 'Earth',
        url: '',
      },
      location: {
        name: 'Earth',
        url: '',
      },
      image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
      episode: [],
      url: '',
      created: '',
    },
    {
      id: 2,
      name: 'Morty Smith',
      status: 'Alive',
      species: 'Human',
      type: '',
      gender: 'Male',
      origin: {
        name: 'Earth',
        url: '',
      },
      location: {
        name: 'Earth',
        url: '',
      },
      image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
      episode: [],
      url: '',
      created: '',
    },
  ],
};

describe('ResultsSection', () => {
  test('renders loader when loading', () => {
    render(
      <MemoryRouter>
        <ResultsSection ifNextPage={false} />
      </MemoryRouter>
    );

    expect(screen.getByTestId('loader')).toBeInTheDocument();
  });

  test('renders character results when not loading', () => {
    render(
      <MemoryRouter>
        <ResultsSection ifNextPage={false} />
      </MemoryRouter>
    );

    expect(screen.getByText('Rick Sanchez')).toBeInTheDocument();
    expect(screen.getByText('Morty Smith')).toBeInTheDocument();
  });

  test('matches snapshot when loading', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <ResultsSection ifNextPage={false} />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('matches snapshot with results', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <ResultsSection ifNextPage={false} />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the specified number of cards', () => {
    render(
      <MemoryRouter>
        <ResultsSection ifNextPage={false} />
      </MemoryRouter>
    );

    const cards = screen.getAllByTestId('results-item');
    expect(mockCharacterResponse.results).toBeDefined();
    if (mockCharacterResponse.results)
      expect(cards).toHaveLength(mockCharacterResponse.results.length);
  });
});
