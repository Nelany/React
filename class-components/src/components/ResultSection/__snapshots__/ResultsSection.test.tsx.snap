// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResultsSection > matches snapshot when loading 1`] = `
<DocumentFragment>
  <div
    class="section results-section light"
  >
    <h3>
      Results:
    </h3>
    <div
      class="spinner-container"
      data-testid="loader"
    >
      <img
        alt="Loading..."
        class="spinner"
        src="/spinner.png"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`ResultsSection > matches snapshot with results 1`] = `
<DocumentFragment>
  <div
    class="section results-section light"
  >
    <h3>
      Results:
    </h3>
    <div
      class="results-item light"
      data-testid="results-item"
    >
      <img
        alt="img"
        class="result-img"
        src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
      />
      <div>
        <h2
          class="light"
        >
          Rick Sanchez
        </h2>
        <h4>
          Status: Alive
        </h4>
        <h4>
          Species: Human
        </h4>
        <h4>
          Last known location: Earth
        </h4>
      </div>
    </div>
    <div
      class="results-item light"
      data-testid="results-item"
    >
      <img
        alt="img"
        class="result-img"
        src="https://rickandmortyapi.com/api/character/avatar/2.jpeg"
      />
      <div>
        <h2
          class="light"
        >
          Morty Smith
        </h2>
        <h4>
          Status: Alive
        </h4>
        <h4>
          Species: Human
        </h4>
        <h4>
          Last known location: Earth
        </h4>
      </div>
    </div>
    <div
      class="pagination light"
    >
      <div
        class="pagination__element pagination__button pagination__button-next pagination__button-disabled"
      >
        &lt;
      </div>
      <div
        class="pagination__element pagination__number"
      >
        1
      </div>
      <div
        class="pagination__element pagination__button pagination__button-next pagination__button-disabled"
      >
        &gt;
      </div>
    </div>
  </div>
</DocumentFragment>
`;
